src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+        echildren.push_back(substrSuffix);
+      }
+      Node body = nm->mkNode(AND, echildren);
+      if (!k.isNull())
+      {
+        Node bvl = nm->mkNode(BOUND_VAR_LIST, k);
+        body = nm->mkNode(EXISTS, bvl, body);
+      }
+      // e.g. x in re.++( R1, "AB", R2 ) --->
+      //  exists k.
+      //    0 <= k <= (len(x)-2) ^
+      //    substr( x, k, 2 ) = "AB" ^
+      //    substr( x, 0, k ) in R1 ^
+      //    substr( x, k+2, len(x)-(k+2) ) in R2
+      return returnElim(atom, body, "concat-find");
+    }
+  }
+  return Node::null();
+}
+
+Node RegExpElimination::eliminateStar(Node atom)
+{
+  if (!options::regExpElimAgg())
+  {
+    return Node::null();
+  }
+  // only aggressive rewrites below here
+
+  NodeManager* nm = NodeManager::currentNM();
+  Node x = atom[0];
+  Node lenx = nm->mkNode(STRING_LENGTH, x);
+  Node re = atom[1];
+  // for regular expression star,
+  // if the period is a fixed constant, we can turn it into a bounded
+  // quantifier
+  std::vector<Node> disj;
+  if (re[0].getKind() == REGEXP_UNION)
+  {
+    for (const Node& r : re[0])
+    {
+      disj.push_back(r);
+    }
+  }
+  else
+  {
+    disj.push_back(re[0]);
+  }
+  bool success = true;
+  std::vector<Node> char_constraints;
+  Node index = nm->mkBoundVar(nm->integerType());
+  Node substr_ch = nm->mkNode(STRING_SUBSTR, x, index, d_one);
+  substr_ch = Rewriter::rewrite(substr_ch);
+  // handle the case where it is purely characters
+  for (const Node& r : disj)
+  {
+    Assert(r.getKind() != REGEXP_SIGMA);
+    success = false;
+    // success is true if the constraint
+    if (r.getKind() == STRING_TO_REGEXP)
+    {
+      Node s = r[0];
+      if (s.isConst() && s.getConst<String>().size() == 1)
+      {
+        success = true;
+      }
+    }
+    else if (r.getKind() == REGEXP_RANGE)
+    {
+      success = true;
+    }
+    if (!success)
+    {
+      break;
+    }
+    else
+    {
+      Node regexp_ch = nm->mkNode(STRING_IN_REGEXP, substr_ch, r);
+      regexp_ch = Rewriter::rewrite(regexp_ch);
+      Assert(regexp_ch.getKind() != STRING_IN_REGEXP);
+      char_constraints.push_back(regexp_ch);
+    }
+  }
+  if (success)
+  {
+    Assert(!char_constraints.empty());
+    Node bound = nm->mkNode(
+        AND, nm->mkNode(LEQ, d_zero, index), nm->mkNode(LT, index, lenx));
+    Node conc = char_constraints.size() == 1 ? char_constraints[0]
+                                             : nm->mkNode(OR, char_constraints);
+    Node body = nm->mkNode(OR, bound.negate(), conc);
+    Node bvl = nm->mkNode(BOUND_VAR_LIST, index);
+    Node res = nm->mkNode(FORALL, bvl, body);
+    // e.g.
+    //   x in (re.* (re.union "A" "B" )) --->
+    //     forall k. 0<=k<len(x) => (substr(x,k,1) = "A" OR substr(x,k,1)="B")
+    return returnElim(atom, res, "star-char");
+  }
+  // otherwise, for stars of constant length these are periodic
+  if (disj.size() == 1)
+  {
+    Node r = disj[0];
+    if (r.getKind() == STRING_TO_REGEXP)
```

Comment:
We can generalize this rewrite to also apply in the case `x in ("A" sigma)*` where we'd assert that every even letter is "A" and the string is of even length (i.e. the regexp within the star can be a regex concat of sigmas and constant strings and the sigmas just count towards the length constraints and each constant string adds another substring constraint).

Reply:
Yep, we can add this later (out of scope for this PR though).